### ISSUE TITLE
Local function can use captured variables which are not definitely assigned

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#41631]
In a scenario like the following, we check the local function usage with the call's postconditions already applied. That is, we assume that the call has assigned to `s0`, and then (simplifying) we go on to analyze the usage of `s0` in the local function body. But that ordering is wrong. We need to modify the AbstractFlowPass so that we check the local function usage after the argument expressions have been visited, but before any ref/out arguments have been marked as written.

```cs
void M()
{
    string s0;
    local1(out s0);

    void local1(out string s1)
    {
        s0.ToString(); // should produce an error, but doesn't
        s1 = ""hello"";
    }
}
```

/cc @agocke

**Comment by RikkiGibson** [comment]
Let's also try to address https://github.com/dotnet/roslyn/pull/41597#issuecomment-585488435
